### PR TITLE
SWIFT-1129 Create full stack Swift example project

### PR DESCRIPTION
**Pull request by kmahar**
SWIFT-1129

### CODE REVIEW

**Comment by kmahar** [comment]
@am-MongoDB I don't seem to be able to formally request you as a reviewer due to repo permissions, but let me know if GitHub will still let you leave comments or not.

File: Examples/FullStackSwiftExample/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+This directory contains a full-stack Swift example application that supports managing a list of kittens via typical CRUD operations. The backend server is written using Vapor and the MongoDB Swift driver, which lives in the [Backend](./Backend) directory. The frontend is an iOS application as defined in [iOSApp](./iOSApp), which communicates with the backend via HTTP. 
+
+The same `Codable` data model types are shared between the frontend and backend, and so they are defined in their own `Models` SwiftPM package.
+
+The backend portion of the application requires Swift 5.5.2+ and MongoDB 3.6+. It will run on Linux as well as macOS 12+. The frontend iOS portion requires Xcode 13.2+.
+
+For more details on each component, please review the corresponding README files: [backend](./Backend/README.md), [frontend](./iOSApp/README.md), [models](./Models/README.md).
+
+## Building and Running the Application
```
[the remark this answers is not in the document]
Comment (isabelatkinson) [reply Examples/FullStackSwiftExample/README.md:11]: all works for me 👍 I've never used the iOS simulator before, very cool!

File: Examples/FullStackSwiftExample/iOSApp/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Kittens iOS Frontend
```
Comment (patrickfreed) [inline Examples/FullStackSwiftExample/iOSApp/README.md:1]: I think it could be cool to include some screenshots here and maybe in the top level readme too

**Review by patrickfreed — COMMENTED** [review]
Looks great! I just have a few minor comments/questions

File: Examples/FullStackSwiftExample/iOSApp/README.md (hunk 2 above)
Comment (kmahar) [reply Examples/FullStackSwiftExample/iOSApp/README.md:1]: good idea, added screenshots to both

**Review by isabelatkinson — APPROVED** [review]
lgtm!

**Review by patrickfreed — APPROVED** [review]
LGTM!

File: Examples/FullStackSwiftExample/iOSApp/Kittens/ViewModels/AddKittenViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        let kitten = Kitten(name: name, color: color, favoriteFood: favoriteFood)
+        try await HTTP.post(url: HTTP.baseURL, body: kitten)
+    }
+
+    /// Adds a new kitten and then runs the provided completion handler.
+    func addAction(completion: @escaping () -> Void) {
+        runInTask(completion: completion) {
+            try await self.addKitten()
+        }
+    }
```
Comment (am-MongoDB) [inline Examples/FullStackSwiftExample/iOSApp/Kittens/ViewModels/AddKittenViewModel.swift:24]: I'd handle the async stuff in the Views
```suggestion
    /// Sends a request to add a new kitten to the backend.
   func addKitten() async throws {
        let kitten = Kitten(name: name, color: color, favoriteFood: favoriteFood)
        try await HTTP.post(url: HTTP.baseURL, body: kitten)
    }
```

File: Examples/FullStackSwiftExample/iOSApp/Kittens/ViewModels/ViewUpdateDeleteKittenViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    /// The initial kitten this view is created with.
+    let kitten: Kitten
+
+    init(currentKitten: Kitten) {
+        self.kitten = currentKitten
+        self.favoriteFood = currentKitten.favoriteFood
+    }
+
+    /// Sends a request to update this kitten to the backend.
+    private func updateKitten() async throws {
```
Comment (am-MongoDB) [inline Examples/FullStackSwiftExample/iOSApp/Kittens/ViewModels/ViewUpdateDeleteKittenViewModel.swift:17]:
```suggestion
   func updateKitten() async throws {
```

File: Examples/FullStackSwiftExample/iOSApp/Kittens/Views/AddKitten.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+                viewModel.addAction {
+                    presentationMode.wrappedValue.dismiss()
+                }
+            }
+            Button("Cancel", role: .cancel) {
+                presentationMode.wrappedValue.dismiss()
+            }
+        }.buttonStyle(.bordered)
+    }
+}
```
Comment (am-MongoDB) [inline Examples/FullStackSwiftExample/iOSApp/Kittens/Views/AddKitten.swift:38]:
```suggestion
import Models
import SwiftUI

/// View to support adding a new kitten.
struct AddKitten: View {
    /// Model for the data in this view.
    @ObservedObject var viewModel: AddKittenViewModel
    /// Presentation mode environment key. This is used to enable the view to dismiss itself on button presses.
    @Environment(\.presentationMode) private var presentationMode
    
    @State private var errorMessage: String?
    @State private var busy = false

    var body: some View {
        NavigationView {
            ZStack {
                VStack {
                    Form {
                        TextField("Name", text: $viewModel.name)
                        TextField("Color", text: $viewModel.color)
                        Picker("Favorite Food", selection: $viewModel.favoriteFood) {
                            ForEach(CatFood.allCases) { food in
                                Text(food.rawValue.capitalized)
                            }
                        }
                    }
                    if let errorMessage = errorMessage {
                        Text(errorMessage)
                            .foregroundColor(.red)
                    }
                    HStack {
                        Button("Add Kitten") {
                            addKitten()
                        }
                        .disabled(viewModel.name == "")
                        .buttonStyle(.borderedProminent)
                        Button("Cancel", role: .cancel) {
                            presentationMode.wrappedValue.dismiss()
                        }
                        .buttonStyle(.bordered)
                    }
                }
                if busy {
                    ProgressView()
                }
            }
        }
    }
    
    private func addKitten() {
        errorMessage = nil
        busy = true
        Task {
            do {
                try await viewModel.addKitten()
                presentationMode.wrappedValue.dismiss()
            } catch {
                errorMessage = "Failed to add kitten: \(error.localizedDescription)"
                busy = false
            }
        }
    }
}

struct AddKitten_Previews: PreviewProvider {
    static var previews: some View {
        AddKitten(viewModel: AddKittenViewModel())
    }
}
```

File: Examples/FullStackSwiftExample/iOSApp/Kittens/Views/KittenList.swift [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            // When the view appears, retrieve an updated list of kittens.
+            runInTask {
+                try await viewModel.fetchKittens()
+            }
+        }
+        Button("Add Kitten") {
+            modalType = .add
+        }.buttonStyle(.bordered)
+    }
+}
```
Comment (am-MongoDB) [inline Examples/FullStackSwiftExample/iOSApp/Kittens/Views/KittenList.swift:73]:
```suggestion
import Models
import SwiftUI

/// Main view displaying a list of kittens.
struct KittenList: View {
    /// Model for the data in this view.
    @StateObject private var viewModel = KittenListViewModel()
    
    @State private var showingAddModal = false
    @State private var busy = false
    @State private var errorMessage: String?
    
    var body: some View {
        NavigationView {
            ZStack {
                VStack {
                    if let errorMessage = errorMessage {
                        Text(errorMessage)
                            .foregroundColor(.red)
                    }
                    List {
                        ForEach(viewModel.kittens) { kitten in
                            // Each element in the list is a button that, if clicked, will open the view/update/delete view for
                            // the corresponding kitten.
                            NavigationLink(destination: ViewUpdateDeleteKitten(viewModel: ViewUpdateDeleteKittenViewModel(currentKitten: kitten))) {
                                Text(kitten.name)
                                    .font(.title3)
                            }
                        }
                    }
                    // Pull to refresh
                    .refreshable(action: { fetchKittens () })
                    Button("Add Kitten") {
                        showingAddModal.toggle()
                    }
                    .buttonStyle(.borderedProminent)
                }
                if busy {
                    ProgressView()
                }
            }
            .sheet(
                isPresented: $showingAddModal,
                onDismiss: {
                    // On dismiss, retrieve an updated list of kittens.
                    fetchKittens()
                }
            ) {
                AddKitten(viewModel: AddKittenViewModel())
            }
            // When the view appears, retrieve an updated list of kittens.
            .onAppear(perform: fetchKittens)
            .navigationBarTitle("Kittens 🐱🐱", displayMode: .inline)
        }
        .navigationViewStyle(StackNavigationViewStyle())
    }
    
    private func fetchKittens() {
        busy = true
        errorMessage = nil
        Task {
            do {
                try await viewModel.fetchKittens()
                busy = false
            } catch {
                busy = false
                errorMessage = "Failed to fetch list of kittens: \(error.localizedDescription)"
            }
        }
    }
}

struct KittenList_Previews: PreviewProvider {
    static var previews: some View {
        KittenList()
    }
}
```

**Review by am-MongoDB — CHANGES_REQUESTED** [review]

File: Examples/FullStackSwiftExample/Models/Package.swift [hunk 7]
```diff
@@ -0,0 +1,29 @@
+// swift-tools-version:5.5
+
+import PackageDescription
+
+let package = Package(
+    name: "models",
+    platforms: [
+        .macOS(.v10_14),
+        .iOS(.v11)
```
[the remark this answers is not in the document]
Comment (kmahar) [reply Examples/FullStackSwiftExample/Models/Package.swift:9]: ah thanks! I updated the min iOS version in swift-bson a few days ago but forgot to update here.

**Pull request merged**